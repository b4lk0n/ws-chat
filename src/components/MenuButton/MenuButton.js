import React, { useState, useCallback, useRef } from 'react';
import cx from 'classnames';
import { string, array, func, any } from 'prop-types';
import { Button } from '../Button';
import { useOnClickOutside, useOnKeyDown } from '../../hooks';
import styles from './MenuButton.module.css';

export const MenuButton = ({ label, items, onChange, value, getItemLabel, getItemKey, className, ...props }) => {
  const [menuOpen, setMenuOpen] = useState(false);
  const toggleMenu = useCallback(() => setMenuOpen(!menuOpen), [menuOpen]);
  const closeMenu = useCallback(() => setMenuOpen(false), []);
  const handleItemClick = useCallback(
    item => () => {
      closeMenu();
      onChange(item);
    },
    [onChange]
  );

  const containerRef = useRef();

  useOnClickOutside(containerRef, closeMenu);
  useOnKeyDown('Escape', closeMenu);

  return (
    <div
      ref={containerRef}
      className={cx(
        styles.root,
        {
          [styles.menuOpen]: menuOpen,
        },
        className
      )}
      data-testid={'menu-button'}
    >
      <Button
        {...props}
        label={value ? getItemLabel(value) : label}
        onClick={toggleMenu}
        data-testid={'menu-button-anchor'}
      />

      <div className={styles.menu} data-testid={'menu-button-list'}>
        {items.map(item => (
          <Button
            key={getItemKey(item)}
            label={getItemLabel(item)}
            onClick={handleItemClick(item)}
            className={styles.menuItem}
            data-testid={'menu-item'}
          />
        ))}
      </div>
    </div>
  );
};
MenuButton.propTypes = {
  label: string.isRequired,
  items: array.isRequired,
  onChange: func.isRequired,
  value: any,
  getItemLabel: func,
  getItemKey: func,
  className: string,
};
MenuButton.defaultProps = {
  getItemLabel: item => item.name,
  getItemKey: item => item.id,
};
