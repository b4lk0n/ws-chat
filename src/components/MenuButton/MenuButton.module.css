.root {
  display: inline-block;
  position: relative;
}

.menu {
  position: absolute;
  width: min-content;
  opacity: 0;
  visibility: hidden;
  background-color: var(--surface);
  border: 1px solid var(--divider);
  border-radius: 4px;
  box-shadow: 0px 6px 10px rgba(0, 0, 0, 0.14), 0px 1px 18px rgba(0, 0, 0, 0.12), 0px 3px 5px rgba(0, 0, 0, 0.2);
  transition: opacity 0.1s ease-in-out, visibility 0.1s ease-in-out;
}

.menuOpen .menu {
  opacity: 1;
  visibility: visible;
}

.menuItem {
  border-radius: 0;
  width: 100%;
}
