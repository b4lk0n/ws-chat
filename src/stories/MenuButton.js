import React, { useState, useCallback } from 'react';
import { withKnobs } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { MenuButton } from '../components';

export default {
  title: 'MenuButton',
  decorators: [withKnobs],
};

const items = [
  { type: 'text', name: 'Text' },
  { type: 'json', name: 'JSON' },
  { type: 'asdas', name: 'application/json' },
];

export const Default = () => {
  const [value, setValue] = useState(null);
  const getItemLabel = useCallback(item => item.name, []);
  const getItemKey = useCallback(item => item.type, []);

  return (
    <MenuButton
      variant={'outlined'}
      label={'Select type'}
      items={items}
      value={value}
      onChange={setValue}
      getItemLabel={getItemLabel}
      getItemKey={getItemKey}
    />
  );
};
