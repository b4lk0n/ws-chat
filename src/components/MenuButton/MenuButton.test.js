import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MenuButton } from './MenuButton';

const items = [
  { type: 'text/plain', name: 'Text' },
  { type: 'application/json', name: 'JSON' },
];

describe('MenuButton', () => {
  it('should render closed by default', () => {
    const { getByTestId } = render(
      <MenuButton items={items} label={'Select'} onChange={jest.fn()} getItemKey={item => item.type} />
    );

    const menuButton = getByTestId('menu-button');
    expect(menuButton).toBeInTheDocument();
    expect(menuButton).not.toHaveClass('menuOpen');
    expect(menuButton).toHaveTextContent('Select');
    expect(menuButton).toMatchSnapshot();
  });

  it('should open menu by clicking on a button', () => {
    const { getByTestId } = render(
      <MenuButton items={items} label={'Select'} onChange={jest.fn()} getItemKey={item => item.type} />
    );

    const menuButton = getByTestId('menu-button');

    expect(menuButton).not.toHaveClass('menuOpen');
    fireEvent.click(getByTestId('menu-button-anchor'));
    expect(menuButton).toHaveClass('menuOpen');
    fireEvent.click(document);
    expect(menuButton).not.toHaveClass('menuOpen');
  });

  it('should select an item by clicking on it', () => {
    const changeHandler = jest.fn();

    const { container } = render(
      <MenuButton items={items} label={'Select'} onChange={changeHandler} getItemKey={item => item.type} />
    );
    expect(container.firstChild.firstChild).toHaveTextContent('Select');
    fireEvent.click(container.firstChild.firstChild);
    expect(container.firstChild).toHaveClass('menuOpen');
    fireEvent.click(container.querySelector('.menuItem'));
    expect(changeHandler).toBeCalledTimes(1);
    expect(changeHandler).toBeCalledWith(items[0]);
    expect(container.firstChild).not.toHaveClass('menuOpen');
  });

  it('should close menu by Escape', () => {
    const { container } = render(
      <MenuButton items={items} label={'Select'} onChange={jest.fn()} getItemKey={item => item.type} />
    );

    expect(container.firstChild).not.toHaveClass('menuOpen');
    fireEvent.click(container.firstChild.firstChild);
    expect(container.firstChild).toHaveClass('menuOpen');
    fireEvent.keyDown(document, { key: 'Escape' });
    expect(container.firstChild).not.toHaveClass('menuOpen');
  });
});
