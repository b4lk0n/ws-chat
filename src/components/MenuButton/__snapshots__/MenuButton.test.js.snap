// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuButton should render closed by default 1`] = `
<div
  class="root"
>
  <button
    class="root undefined"
  >
    <span
      class="root subtitle label"
    >
      Select
    </span>
  </button>
  <div
    class="menu"
  >
    <button
      class="root undefined menuItem"
    >
      <span
        class="root subtitle label"
      >
        Text
      </span>
    </button>
    <button
      class="root undefined menuItem"
    >
      <span
        class="root subtitle label"
      >
        JSON
      </span>
    </button>
  </div>
</div>
`;
